tools/gfx/renderer-shared.cpp

```diff
@@ -325,6 +331,118 @@ void PipelineStateBase::initializeBase(const PipelineStateDesc& inDesc)
     }
 }
 
+void updateCacheEntry(ISlangMutableFileSystem* fileSystem, slang::IBlob* compiledCode, String shaderFilename, slang::Checksum ASTHash)
+{
+    auto hashSize = sizeof(slang::Checksum);
+
+    auto bufferSize = hashSize + compiledCode->getBufferSize();
+    List<uint8_t> contents;
+    contents.setCount(bufferSize);
+    uint8_t* buffer = contents.begin();
+    memcpy(buffer, (void*)ASTHash.checksum, hashSize);
+    memcpy(buffer + hashSize, (void*)compiledCode->getBufferPointer(), compiledCode->getBufferSize());
+    fileSystem->saveFile(shaderFilename.getBuffer(), buffer, bufferSize);
+}
+
+Result RendererBase::getEntryPointCodeFromShaderCache(
+    slang::IComponentType* program,
+    SlangInt entryPointIndex,
+    SlangInt targetIndex,
+    slang::IBlob** outCode,
+    slang::IBlob** outDiagnostics)
+{
+    // TODO: Need a way in filesystem to query both file size and file creation time, if cache size exceeds
+    // specified maximum size (in bytes or files) then delete oldest files - cache eviction policy
+
+    // Immediately call getEntryPointCode if no shader cache was provided on initialization
+    if (!shaderCacheFileSystem)
+    {
+        return program->getEntryPointCode(entryPointIndex, targetIndex, outCode, outDiagnostics);
+    }
+
+    // Produce a string which we can use to query the shader cache by combining two separate hashes which
+    // together comprise all the compilation arguments for this program.
+    ComPtr<slang::ISession> session;
+    getSlangSession(session.writeRef());
+
+    slang::Checksum linkageHash;
+    slang::Checksum programHash;
+    session->computeDependencyBasedHash(&linkageHash);
+    program->computeDependencyBasedHash(entryPointIndex, targetIndex, &programHash);
+
+    StringBuilder shaderFilename = checksumToString(linkageHash);
+    shaderFilename.append(checksumToString(programHash));
```

Comment:
This is not correct. Should combine `programHash` and `linkageHash` for the final fileName. Just run MD5 again on these two hashes. Consider adding a new helper function: `Checksum combineChecksum(const Checksum& v0, const Checksum& v1)

Reply:
Concatenating the hashes makes the file name 64 characters long, which isn't necessary. Instead write a `Checksum combineChecksum(const Checksum& c0, const Checksum c1)`, and then do:
```
auto combinedChecksum = combineChecksum(linkageHash, programHash);
auto shaderFilename = checksumToString(combinedChecksum);
```

Reply:
Done in the most recent push after moving `slang-md5` back to `source/core` because it seems `renderer-shared.cpp` throws unresolved external errors at me if it includes that file when it's located in `source/compiler-core`